src/Components/Web.Extensions/src/InputFile/InputFile.cs

```diff
@@ -0,0 +1,109 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Components.Rendering;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.JSInterop;
+
+namespace Microsoft.AspNetCore.Components.Web.Extensions
+{
+    /// <summary>
+    /// A component that wraps the HTML file input element and exposes a <see cref="Stream"/> for each file's contents.
+    /// </summary>
+    public class InputFile : ComponentBase, IInputFileJsCallbacks, IDisposable
+    {
+        private ElementReference _inputFileElement;
+
+        private IJSRuntime _jsRuntime = default!;
+
+        private IJSUnmarshalledRuntime? _jsUnmarshalledRuntime;
+
+        private InputFileJsCallbacksRelay? _jsCallbacksRelay;
+
+        [Inject]
+        private IServiceProvider ServiceProvider { get; set; } = default!;
+
+        /// <summary>
+        /// Gets or sets the event callback that will be invoked when the collection of selected files changes.
+        /// </summary>
+        [Parameter]
+        public EventCallback<IFileListEntry[]> OnChange { get; set; }
+
+        /// <summary>
+        /// Gets or sets the maximum message size for file data sent over a circuit.
+        /// This only has an effect when using Blazor Server.
+        /// </summary>
+        [Parameter]
+        public int MaxMessageSize { get; set; } = 20 * 1024; // SignalR limit is 32K.
+
+        /// <summary>
+        /// Gets or sets the maximum internal buffer size for file data sent over a circuit.
+        /// This only has an effect when using Blazor Server.
+        /// </summary>
+        [Parameter]
+        public int MaxBufferSize { get; set; } = 1024 * 1024;
+
+        /// <summary>
+        /// Gets or sets a collection of additional attributes that will be applied to the input element.
+        /// </summary>
+        [Parameter(CaptureUnmatchedValues = true)]
+        public IDictionary<string, object>? AdditionalAttributes { get; set; }
+
+        protected override void OnInitialized()
+        {
+            _jsRuntime = ServiceProvider.GetRequiredService<IJSRuntime>();
+            _jsUnmarshalledRuntime = ServiceProvider.GetService<IJSUnmarshalledRuntime>();
```

Comment:
For the 5.0 release, I'd personally prefer *not* to have a DI registration for `IJSUnmarshalledRuntime` explicitly, because it's a super advanced case that people should hopefully only start using when they get to it via docs. If we make it too easy to get an `IJSUnmarshalledRuntime`, people may tend to use it more often than they should, which is fraught with difficulties, plus breaks the whole model of components being portable across Blazor WebAssembly and Blazor Server.

So, for 5.0 at least, I suggest that to get an `IJSUnmarshalledRuntime` people should have to do this:

```
@inject IJSRuntime JSRuntime

@code {
    void SomeMethod()
    {
        if (JSRuntime is IJSUnmarshalledRuntime unmarshalledRuntime)
        {
            // Do stuff with unmarshalledRuntime
        }
        else
        {
            // Do stuff with JSRuntime
        }
    }
}
```

If we follow this approach then another benefit is that `InputFile` doesn't need to inject the whole `IServiceProvider`. It can just inject the `IJSRuntime` and then attempts a runtime cast.